controllers/service_monitor.go

```diff
@@ -0,0 +1,171 @@
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/kubernetes"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+	ctrllog "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+type ServiceMonitor struct {
+	Enable bool
+	Mondoo v1alpha1.MondooAuditConfig
+}
+
+func (s *ServiceMonitor) declareServiceMonitor(ctx context.Context, clt client.Client, scheme *runtime.Scheme) (ctrl.Result, error) {
+	log := ctrllog.FromContext(ctx)
+
+	found := &monitoringv1.ServiceMonitor{}
+	err := clt.Get(ctx, types.NamespacedName{Name: s.Mondoo.Name + "-metrics-monitor", Namespace: s.Mondoo.Namespace}, found)
+	if err != nil && errors.IsNotFound(err) {
+
+		declared := s.serviceMonitorForMondoo(&s.Mondoo)
+		if err := ctrl.SetControllerReference(&s.Mondoo, declared, scheme); err != nil {
+			log.Error(err, "Failed to set ControllerReference", "ServiceMonitor.Namespace", declared.Namespace, "ServiceMonitor.Name", declared.Name)
+			return ctrl.Result{}, err
+		}
+
+		err := clt.Create(ctx, declared)
+		if err != nil {
+			log.Error(err, "Failed to create new ServiceMonitor", "ServiceMonitor.Namespace", declared.Namespace, "ServiceMonitor.Name", declared.Name)
+			return ctrl.Result{}, err
+		}
+
+		return ctrl.Result{Requeue: true}, err
+
+	} else if err != nil {
+		log.Error(err, "Failed to get ServiceMonitor")
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (s *ServiceMonitor) serviceMonitorForMondoo(m *v1alpha1.MondooAuditConfig) *monitoringv1.ServiceMonitor {
+	ls := labelsForMondoo(m.Name)
+	dep := &monitoringv1.ServiceMonitor{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      m.Name + "-metrics-monitor",
+			Namespace: m.Namespace,
+			Labels:    ls,
+		},
+		Spec: monitoringv1.ServiceMonitorSpec{
+			Endpoints: []monitoringv1.Endpoint{
+				{
+					Path:   "/metrics",
+					Port:   "https",
+					Scheme: "https",
+					TLSConfig: &monitoringv1.TLSConfig{
+						SafeTLSConfig: monitoringv1.SafeTLSConfig{
+							InsecureSkipVerify: true,
+						},
+					},
+					BearerTokenFile: "/var/run/secrets/kubernetes.io/serviceaccount/token",
+				},
+			},
+			Selector: metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					"control-plane": "controller-manager",
+				},
+			},
+		},
+	}
+	return dep
+}
+func (s *ServiceMonitor) Reconcile(ctx context.Context, clt client.Client, scheme *runtime.Scheme) (ctrl.Result, error) {
+
+	found, err := VerifyAPI(monitoringv1.SchemeGroupVersion.Group, monitoringv1.SchemeGroupVersion.Version, ctx)
+
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	if found {
+		if s.Enable {
+			result, err := s.declareServiceMonitor(ctx, clt, scheme)
+			if err != nil || result.Requeue {
+				return result, err
+			}
+		} else {
+			s.down(ctx, clt)
+		}
+		return ctrl.Result{}, nil
+	}
+	return ctrl.Result{}, nil
+}
```

Comment:
We should bubble up the fact that someone has tried to enable metrics, but the metrics CRDs missing as a Condition into the MondooAuditConfig.Status.Conditions[] (this is written up in https://github.com/mondoohq/mondoo-operator/issues/141 ).
This would allow a flow like:
```suggestion
	if s.Enable {
		if !found {
			// Update MondooAuditConfig.Status to communicate why metrics couldn't be configured
			return reconcile.Result{}, nil
		} else {
		      // Update MondoAuditConfig.Status to say that we are not blocked on prometheus CRDs not being installed (in case we were previously in an error state we need to be sure to clear it here)
		  }
		result, err := s.declareServiceMonitor(ctx, clt, scheme)
		if err != nil || result.Requeue {
			return result, err
		}
	} else {
		if found {
			s.down(ctx, clt)
		}
	}
```

Reply:
@joelddiaz Shall we add this as requirement for #141